tests/Aspire.Hosting.Tests/ResourceNotificationTests.cs

```diff
@@ -147,6 +148,149 @@ await foreach (var item in notificationService.WatchAsync(cancellation))
+    {
+        var resource1 = new CustomResource("myResource1");
+
+        var notificationService = new ResourceNotificationService(new NullLogger<ResourceNotificationService>(), new TestHostApplicationLifetime());
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+        var waitTask = notificationService.WaitForResourceAsync("myResource1", "SomeState", cts.Token);
+
+        await notificationService.PublishUpdateAsync(resource1, snapshot => snapshot with { State = "SomeState" });
+        await waitTask;
+
+        Assert.True(waitTask.IsCompletedSuccessfully);
+    }
+
+    [Fact]
+    public async Task WaitingOnResourceReturnsWhenResourceReachesTargetStateWithDifferentCasing()
+    {
+        var resource1 = new CustomResource("myResource1");
+
+        var notificationService = new ResourceNotificationService(new NullLogger<ResourceNotificationService>(), new TestHostApplicationLifetime());
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+        var waitTask = notificationService.WaitForResourceAsync("MYreSouRCe1", "sOmeSTAtE", cts.Token);
+
+        await notificationService.PublishUpdateAsync(resource1, snapshot => snapshot with { State = "SomeState" });
+        await waitTask;
+
+        Assert.True(waitTask.IsCompletedSuccessfully);
+    }
+
+    [Fact]
+    public async Task WaitingOnResourceReturnsImmediatelyWhenResourceIsInTargetStateAlready()
+    {
+        var resource1 = new CustomResource("myResource1");
+
+        var notificationService = new ResourceNotificationService(new NullLogger<ResourceNotificationService>(), new TestHostApplicationLifetime());
+
+        // Publish the state update first
+        await notificationService.PublishUpdateAsync(resource1, snapshot => snapshot with { State = "SomeState" });
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+        var waitTask = notificationService.WaitForResourceAsync("myResource1", "SomeState", cts.Token);
+
+        Assert.True(waitTask.IsCompletedSuccessfully);
+    }
+
+    [Fact]
+    public async Task WaitingOnResourceReturnsWhenResourceReachesRunningStateIfNoTargetStateSupplied()
+    {
+        var resource1 = new CustomResource("myResource1");
+
+        var notificationService = new ResourceNotificationService(new NullLogger<ResourceNotificationService>(), new TestHostApplicationLifetime());
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+        var waitTask = notificationService.WaitForResourceAsync("myResource1", targetState: null, cancellationToken: cts.Token);
+
+        await notificationService.PublishUpdateAsync(resource1, snapshot => snapshot with { State = KnownResourceStates.Running });
+        await waitTask;
+
+        Assert.True(waitTask.IsCompletedSuccessfully);
+    }
+
+    [Fact]
+    public async Task WaitingOnResourceReturnsCorrectStateWhenResourceReachesOneOfTargetStatesBeforeCancellation()
```

Comment:
`ReachesOneOfTargetStatesBeforeCancellation` - what is this checking for exactly? When should the cancellation be happening for the test? Relying only on the timeout would mean that the cancellation could happen before/after the publish, or before the new state gets read, IIUC.

Reply:
It's simply testing the happy-path case of passing multiple target states and ensuring the task returns with the first state that was reached. The timeout CTS is only there to ensure the test doesn't hang if there's an issue.